### PR TITLE
feat: Support pretokenized

### PR DESCRIPTION
**Pull request by kmehant**

### CODE REVIEW

File: tuning/utils/preprocessing_utils.py [hunk 1]
```diff
@@ -29,8 +29,48 @@
 logger = logging.get_logger("sft_trainer_preprocessing")
 
 
+def _is_pretokenized_dataset(dataset: Dataset):
```
Comment (fabianlim) [inline tuning/utils/preprocessing_utils.py:32]: i suggest move this to inside `validate_data_args` and put a comment that his is how SFT checks for tokenized internally.

File: tuning/utils/preprocessing_utils.py [hunk 2]
```diff
@@ -29,8 +29,48 @@
...
+    return ("input_ids" in dataset.column_names) and ("labels" in dataset.column_names)
+
+
 def validate_data_args(data_args: configs.DataArguments, packing: bool):
 
+    # validation for pretokenized datasets case
```
Comment (fabianlim) [inline tuning/utils/preprocessing_utils.py:38]: this kid of checks can be hard to parse, i suggest to document in plain english above.

File: tuning/utils/preprocessing_utils.py (hunk 1 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: @fabianlim done

File: tuning/utils/preprocessing_utils.py (hunk 2 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: @fabianlim documented them whereever possible thanks

File: tuning/utils/preprocessing_utils.py [hunk 3]
```diff
@@ -86,11 +149,22 @@ def get_data_collator(
...
+            training data arguments
 
     Returns:
         Callable
             Callable collator to be leveraged by the trainer.
     """
+    if data_args.is_pretokenized:
+        return DataCollatorForSeq2Seq(
+            tokenizer=tokenizer,
+            padding="max_length",
```
Comment (fabianlim) [inline tuning/utils/preprocessing_utils.py:164]: @kmehant can we handle the datacollator kwargs more carefully, because this collator calls `pad()` underneath, and there are various options 

https://github.com/huggingface/transformers/blob/ef177a5e1cdf0ca53e24e6d76e813198f7300dc4/src/transformers/tokenization_utils_base.py#L3418-L3429

If we decide for the user what options, then we need to chose the best one, for example we probably do not want to chose `padding='max_length'` aor `padding=None`

Comment (fabianlim) [reply tuning/utils/preprocessing_utils.py:164]: And if there are defaults, then we can see below https://github.com/foundation-model-stack/fms-hf-tuning/pull/272/files#diff-0949af2da3611757d2552841654a34f1fe0a349ee114ce6342b9e0ade3eb13a6R180 how they are handled in `configs`

Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: Thanks, updated!

Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: > On the fence on "pt" or configs.PADDING_RETURN_TENSORS


Avoided this since collator should mostly return tensors rather python lists.

File: tuning/utils/preprocessing_utils.py [hunk 4]
```diff
@@ -29,8 +29,55 @@
 logger = logging.get_logger("sft_trainer_preprocessing")
 
 
+# check if the provided dataset is pretokenized or not
+# the check is taken from trl
+# https://github.com/huggingface/trl/blob/ddf4c8dc3ecf6d9ee2b24f94c62182ffd682c808/trl/trainer/sft_trainer.py#L498-L509
+def is_pretokenized_dataset(data_path: str):
+    if data_path:
+        # load one sample from the dataset in order to inspect columns
+        dataset = datasets.load_dataset("json", data_files=data_path, split="train[:1]")
```
Comment (alex-jw-brooks) [inline tuning/utils/preprocessing_utils.py:38]: Can you add graceful handling for if the dataset is empty?

File: tuning/utils/preprocessing_utils.py [hunk 5]
```diff
@@ -86,11 +133,24 @@ def get_data_collator(
...
             Callable collator to be leveraged by the trainer.
     """
+    is_train_data_pretokenized = is_pretokenized_dataset(data_path=training_data_path)
+
+    if is_train_data_pretokenized:
+        return DataCollatorForSeq2Seq(
+            tokenizer=tokenizer,
+            padding=configs.PADDING_STRATEGY_LONGEST,
+            max_length=max_sequence_length,
+            return_tensors="pt",
```
Comment (alex-jw-brooks) [inline tuning/utils/preprocessing_utils.py:152]: nit - I'd suggest not setting `return_tensors="pt"` directly here, since it is the default, and we don't set it when initializing other collators

File: tuning/utils/preprocessing_utils.py [hunk 6]
```diff
@@ -29,8 +29,55 @@
...
+
+    # if the provided train dataset is pretokenized
+    # however user provides formatting flags, error out
+    if is_train_data_pretokenized:
+        if (
+            data_args.response_template
+            or data_args.data_formatter_template
+            or data_args.dataset_text_field
+        ):
+            raise ValueError(
```
Comment (alex-jw-brooks) [inline tuning/utils/preprocessing_utils.py:62]: Can you change this to a warning instead of an error? IMO this and packing=True should be handled the same way

**Review by alex-jw-brooks — CHANGES_REQUESTED** [review]

File: tuning/utils/preprocessing_utils.py [hunk 7]
```diff
@@ -68,13 +115,13 @@ def validate_data_args(data_args: configs.DataArguments, packing: bool):
     # TODO(s) In future seupport two more formats:
     # 1. Allow no response template, and JSON with input/output fields and mask input
 
-    # 2. Allow pretokenized Dataset besides JSON.
-
 
 def get_data_collator(
```
Comment (Ssukriti) [inline tuning/utils/preprocessing_utils.py:119]: unit tests need to be added for updates. see https://github.com/foundation-model-stack/fms-hf-tuning/blob/6e836d14522de9092457cf63386ae6b20eaa2a2d/tests/utils/test_preprocessing_utils.py#L129 for reference. 
Probably after rebasing, we can re-use same unit test

File: tuning/utils/preprocessing_utils.py [hunk 8]
```diff
@@ -29,8 +29,55 @@
...
+        return ("input_ids" in dataset.column_names) and (
+            "labels" in dataset.column_names
+        )
+    return False
+
+
 def validate_data_args(data_args: configs.DataArguments, packing: bool):
 
+    is_train_data_pretokenized = is_pretokenized_dataset(
+        data_path=data_args.training_data_path
```
Comment (Ssukriti) [inline tuning/utils/preprocessing_utils.py:48]: unit tests need to be added as well

File: tuning/utils/preprocessing_utils.py [hunk 9]
```diff
@@ -129,6 +189,21 @@ def format_dataset(data_args: configs.DataArguments, tokenizer: AutoTokenizer):
             tuple containing train_dataset, eval_dataset and dataset_text_field
     """
     eval_dataset = None
+    is_train_data_pretokenized = is_pretokenized_dataset(
+        data_path=data_args.training_data_path
```
Comment (Ssukriti) [inline tuning/utils/preprocessing_utils.py:193]: unit tests

Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: Added, thanks

File: tuning/utils/preprocessing_utils.py (hunk 7 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: using the same unit test, thanks. 

But if you see some sort of duplicated flow in this function let me know, we will work it out, thanks

File: tuning/utils/preprocessing_utils.py (hunk 6 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: modified thanks @alex-jw-brooks

File: tuning/utils/preprocessing_utils.py (hunk 5 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: ah, right! updated

File: tuning/utils/preprocessing_utils.py (hunk 4 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: nesting it from the same error class `DatasetGenerationError` since there are tests on malformed data already that assert on DatasetGenerationError Thanks. let me know if the message has to be refined. Thanks

File: tuning/utils/preprocessing_utils.py [hunk 10]
```diff
@@ -125,10 +173,7 @@ def get_data_collator(
             )
         # Note that this automatically pads labels with -100
         # TODO check if this is sufficient for preprocessed
-        if (
-            "attention_mask" in formatted_train_dataset.column_names
-            and "labels" in formatted_train_dataset.column_names
-        ):
+        if is_train_data_pretokenized:
```
Comment (Ssukriti) [inline tuning/utils/preprocessing_utils.py:176]: packing is anyway not supported for pretokenized as well, so I just moved it here @kmehant like it was before. The check was doing same thing as your function is_pretokenized. 

padding=True is same as padding = longest . see https://huggingface.co/transformers/v4.9.2/main_classes/data_collator.html

File: tuning/utils/preprocessing_utils.py [hunk 11]
```diff
@@ -33,8 +34,53 @@
...
+
+    # if the provided train dataset is pretokenized
+    # however user provides formatting flags, error out
+    if is_train_data_pretokenized:
+        if (
+            data_args.response_template
+            or data_args.data_formatter_template
+            or data_args.dataset_text_field
+        ):
+            logger.warning(
```
Comment (Ssukriti) [inline tuning/utils/preprocessing_utils.py:65]: should be a valueerror . If response template is provided, we will return completionLMcollator. There is no need to allow a response template, it is an incorrect argument.
I am committing this change

File: tuning/utils/preprocessing_utils.py [hunk 12]
```diff
@@ -33,8 +34,53 @@
...
+        if data_args.validation_data_path and not is_eval_data_pretokenized:
+            raise ValueError(
+                "validation data should be pretokenized to be used \
+                along with pretokenized train data"
+            )
+
+        # packing wont be available for pretokenized datasets in trl library
+        # see: https://github.com/huggingface/trl/issues/1848
+        if packing:
+            logger.warning("packing will not be used when datasets are pretokenized")
```
Comment (Ssukriti) [inline tuning/utils/preprocessing_utils.py:81]: why is this only a warning? it is ok to be warning if code does not perform as well, like performance wont be good etc. We cant have warning when code just wont work. If flag is set to packing , it will be true even when passing to TRL

File: tuning/utils/preprocessing_utils.py [hunk 13]
```diff
@@ -33,12 +34,59 @@
...
+
+
 def validate_data_args(data_args: configs.DataArguments, packing: bool):
 
     assert isinstance(
         data_args.training_data_path, str
     ), "Training data path has to be set and str"
 
+    is_train_data_pretokenized = is_pretokenized_dataset(data_args.training_data_path)
+    is_eval_data_pretokenized = is_pretokenized_dataset(data_args.validation_data_path)
```
Comment (Ssukriti) [inline tuning/utils/preprocessing_utils.py:59]: this is going to return False when validation file not set. later in line 77, it will return valueerror .

validation file is optional from top level API. Only when not None, these checks need to happen. Please add unit test

Comment (Ssukriti) [reply tuning/utils/preprocessing_utils.py:59]: oh I see you added `if data_args.validation_data_path` later on, thats fine then

File: tuning/utils/preprocessing_utils.py (hunk 10 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py:177]: Thanks

File: tuning/utils/preprocessing_utils.py (hunk 11 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: sure @Ssukriti thanks

File: tuning/utils/preprocessing_utils.py (hunk 12 above)
Comment (kmehant) [reply tuning/utils/preprocessing_utils.py]: Agree with you, we should error out

**Comment by Ssukriti** [comment]
tests are still incomplete for both validate_args and format_dataset . I understand its because the JSON is not readily available for tests. You can take a few examples from any JSON here https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tests/data/twitter_complaints_input_output.json and convert it to pretokenized and contribute new JSON with pretokenized. Then it can be easily used in all tests

**Review by Ssukriti — APPROVED** [review]

**Review by alex-jw-brooks — APPROVED** [review]
